src/librustc/ty/subst.rs

```diff
@@ -305,27 +297,22 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
                        target_substs: &Substs<'tcx>)
                        -> &'tcx Substs<'tcx> {
         let defs = tcx.lookup_generics(source_ancestor);
-        tcx.mk_substs(Substs {
-            params: target_substs.params.iter()
-                .chain(&self.params[defs.own_count()..]).cloned().collect()
-        })
+        Substs::new(tcx, target_substs.iter().chain(&self[defs.own_count()..]).cloned())
     }
 }
 
 impl<'tcx> TypeFoldable<'tcx> for &'tcx Substs<'tcx> {
     fn super_fold_with<'gcx: 'tcx, F: TypeFolder<'gcx, 'tcx>>(&self, folder: &mut F) -> Self {
-        let params = self.params.iter().map(|k| k.fold_with(folder)).collect();
-        folder.tcx().mk_substs(Substs {
-            params: params
-        })
+        let params = self.iter().map(|k| k.fold_with(folder)).collect();
```

Comment:
Perhaps you can help me think of a better solution -- `Substs::new` takes `IntoIterator<Item=Kind>`, which is what we'd really want here. But the closure in the `map` call here borrows `folder` mutably, and `Substs::new` wants a `TyCtxt` from the folder, trying to borrow that immutably. And trying to create a simultaneous pair of mutable/immutable borrows obviously creates a borrow error.

The "workaround" I decided on was to allocate here (it has to happen somewhere before interning), and then call the intern function directly instead of going through the iteration-based approach. This should still only allocate the `Vec<Kind>` the once, but its an ugly bypass of the "right" interface.